src/core/server/http/base_path_proxy_server.ts

```diff
@@ -0,0 +1,162 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { RouteOptions, Server } from 'hapi-latest';
+import { Agent as HttpsAgent, ServerOptions as TlsOptions } from 'https';
+import { sample } from 'lodash';
+import { ByteSizeValue } from '../config/schema';
+import { DevConfig } from '../dev';
+import { Logger } from '../logging';
+import { HttpConfig } from './http_config';
+import { createServer, getServerOptions } from './http_tools';
+
+const alphabet = 'abcdefghijklmnopqrztuvwxyz'.split('');
+
+export interface BasePathProxyServerOptions {
+  httpConfig: HttpConfig;
+  devConfig: DevConfig;
+  isKibanaPath: (path: string) => boolean;
+  blockUntil?: () => Promise<void>;
```

Comment:
Just a general question here, why is `log` not considered one of the options, but instead a separate argument to the constructor? httpconfig and devconfig are considered options, but not the log? Wondering about the reasoning.

Reply:
That's a good question and I agree there is no much logical difference between `log` and the rest of the options here. I usually only put __optional__ parameters into `options` if I have (or plan to have) more than one such parameter. In this particular case I just used pattern similar to what we use in other `*_server.ts` within this folder (log is passed as a first separate argument)... Let me know if you want me to put `log` into `options` as well - I'll do that.

Reply:
I just wanted to say, I don't feel a strong preference.

Reply:
Okay, then I'll leave it as is for now, but will keep it mind so that we can discuss this convention later once we start implementing more core services.